File: vuln.go
// vuln.go
package main

import (
	"fmt"
	"os"
	"os/exec"
)

func runCmd(input string) {
	exec.Command("sh", "-c", input).Run()
}

func readFile(name string) {
	// 路径穿越
	path := "/safe/dir/" + name
	os.ReadFile(path)
}

func main() {
	if len(os.Args) > 1 {
		runCmd(os.Args[1])
		readFile(os.Args[1])
		fmt.Println("done")
	}
}

File: vuln.java
// vuln.java
import java.io.*;
import java.util.*;

public class Vuln {
    public static void runCmd(String input) throws Exception {
        // 命令执行
        Runtime.getRuntime().exec(input);
    }

    public static void readFile(String name) throws Exception {
        // 路径穿越
        File f = new File("/safe/dir/" + name);
        new FileInputStream(f);
    }

    public static void main(String[] args) throws Exception {
        if (args.length > 0) {
            runCmd(args[0]);
            readFile(args[0]);
        }
    }
}